File: 백준/Gold/1726. 로봇/로봇.py
# 북동남서 시계 방향
dx = [-1,0,1,0]
dy = [0,1,0,-1]

# 방향 재계산 동서남북 -> 북동남서 
def get_real_dir(dir):
  if dir == 1: # 동
    return 1
  elif dir == 2: #서
    return 3
  elif dir == 3: #남
    return 2
  elif dir == 4: #북
    return 0

# 입력
m, n = map(int,input().split())
grid = [list(map(int,input().split())) for _ in range(m)]
sx,sy,sd = map(int,input().split())
ex,ey,ed = map(int,input().split())
k = 3 # 이동 가능 거리

# 1-based -> 0-based
sx,sy = sx - 1, sy - 1
ex,ey = ex - 1, ey - 1
# 방향 재할당
sd = get_real_dir(sd)
ed = get_real_dir(ed)

from collections import deque
# 완전 탐색 : bfs
def bfs():
    visited = [[[False]*4 for _ in range(n)] for _ in range(m)]
    queue = deque()
    queue.append((sx, sy, sd, 0))
    visited[sx][sy][sd] = True

    while queue:
        x, y, d, move = queue.popleft()

        if (x, y, d) == (ex, ey, ed):
            return move

        # 좌우 회전
        for nd in [(d-1)%4, (d+1)%4]:
            if not visited[x][y][nd]:
                visited[x][y][nd] = True
                queue.append((x, y, nd, move + 1))

        # 직진 1~3칸
        for dist in range(1, 4):
            nx = x + dx[d]*dist
            ny = y + dy[d]*dist
            if 0 <= nx < m and 0 <= ny < n:
                if grid[nx][ny] == 1:
                    break
                if not visited[nx][ny][d]:
                    visited[nx][ny][d] = True
                    queue.append((nx, ny, d, move + 1))
            else:
                break

print(bfs())

File: 백준/Gold/1726. 로봇/로봇.js
const input = require('fs')
	.readFileSync('/dev/stdin', 'utf-8')
	.trim()
	.split('\n');

// 동 남 서 북
const dx = [0, 1, 0, -1];
const dy = [1, 0, -1, 0];

const dirToIdx = {
	1: 0,
	2: 2,
	3: 1,
	4: 3,
};

const [n, m] = input[0].trim().split(' ').map(Number);
const grid = Array.from({ length: n }, (_, idx) =>
	input[idx + 1].trim().split(' ').map(Number)
);
let [sx, sy, sd] = input[n + 1].trim().split(' ').map(Number);
[sx, sy, sd] = [sx - 1, sy - 1, dirToIdx[sd]];
let [ex, ey, ed] = input[n + 2].trim().split(' ').map(Number);
[ex, ey, ed] = [ex - 1, ey - 1, dirToIdx[ed]];

const visited = Array.from({ length: n }, () =>
	Array.from({ length: m }, () => Array(4).fill(-1))
);

queue = [];
queue.push([sx, sy, sd, 0]);
visited[sx][sy][sd] = 0;

while (queue.length > 0) {
	const [x, y, cd, cnt] = queue.shift();

	if (x === ex && y === ey && cd === ed) {
		console.log(visited[ex][ey][ed]);
		break;
	}

	for (let k = 1; k < 4; k++) {
		const nx = x + k * dx[cd];
		const ny = y + k * dy[cd];

		if (!(nx >= 0 && nx < n && ny >= 0 && ny < m)) break;
		if (grid[nx][ny] === 1) break;
		if (visited[nx][ny][cd] === -1) {
			visited[nx][ny][cd] = cnt + 1;
			queue.push([nx, ny, cd, cnt + 1]);
		}
	}

	// 오른쪽
	let nd = (cd + 1) % 4;
	if (visited[x][y][nd] === -1) {
		visited[x][y][nd] = cnt + 1;
		queue.push([x, y, nd, cnt + 1]);
	}

	nd = (cd + 3) % 4;
	if (visited[x][y][nd] === -1) {
		visited[x][y][nd] = cnt + 1;
		queue.push([x, y, nd, cnt + 1]);
	}
}
